### ISSUE TITLE
Smoother block animations

### ISSUE THREAD

**Issue by AJenbo** [diasurgical/devilutionX#984]
Currently block animations frame skipping is unevenliy distributed leading to some bad jitter during playback. This is controlled from PM_DoGotHit and the resulting behavior is as follows (note in helfire there is no zen combination):
```
-= Diablo =-

Warrior  Normal  : 1, 2, 3, 4, 5, 6
Warrior  bal     : 1, 2, 3, -, 5, 6
Warrior  sta     : 1, 2, 3, -, 5, -
Warrior  har     : 1, -, 3, -, 5, -
Warrior  bal+sta : 1, 2, 3, -, -, 6
Warrior  bal+har : 1, -, 3, -, -, 6
Warrior  sta+har : 1, -, 3, -, -, 6
Warrior  zen     : 1, -, 3, -, -, -

Rogue    Normal  : 1, 2, 3, 4, 5, 6, 7
Rogue    bal     : 1, 2, 3, -, 5, 6, 7
Rogue    sta     : 1, 2, 3, -, 5, -, 7
Rogue    har     : 1, -, 3, -, 5, -, 7
Rogue    bal+sta : 1, 2, 3, -, -, 6, 7
Rogue    bal+har : 1, -, 3, -, -, 6, 7
Rogue    sta+har : 1, -, 3, -, -, 6, 7
Rogue    zen     : 1, -, 3, -, -, -, 7

Sorceror Normal  : 1, 2, 3, 4, 5, 6, 7, 8
Sorceror bal     : 1, 2, 3, -, 5, 6, 7, 8
Sorceror sta     : 1, 2, 3, -, 5, -, 7, 8
Sorceror har     : 1, -, 3, -, 5, -, 7, 8
Sorceror bal+sta : 1, 2, 3, -, -, 6, 7, 8
Sorceror bal+har : 1, -, 3, -, -, 6, 7, 8
Sorceror sta+har : 1, -, 3, -, -, 6, 7, 8
Sorceror zen     : 1, -, 3, -, -, -, 7, 8
```

A better pattern would be the following:
```
Warrior  Normal  : 1, 2, 3, 4, 5, 6
Warrior  bal     : 1, 2, 3, -, 5, 6
Warrior  sta     : 1, -, 3, 4, -, 6
Warrior  har     : 1, -, 3, -, 5, -
Warrior  bal+sta : 1, -, 3, 4, -, 6
Warrior  bal+har : 1, -, 3, -, 5, -
Warrior  sta+har : 1, -, 3, -, 5, -
Warrior  zen     : -, 2, -, -, 5, -

Rogue    Normal  : 1, 2, 3, 4, 5, 6, 7
Rogue    bal     : 1, 2, 3, -, 5, 6, 7
Rogue    sta     : 1, -, 3, 4, 5, -, 7
Rogue    har     : 1, -, 3, -, 5, -, 7
Rogue    bal+sta : 1, -, 3, 4, 5, -, 7
Rogue    bal+har : 1, -, 3, -, 5, -, 7
Rogue    sta+har : 1, -, 3, -, 5, -, 7
Rogue    zen     : 1, -, -, 3, -, -, 7

Sorceror Normal  : 1, 2, 3, 4, 5, 6, 7, 8
Sorceror bal     : 1, 2, 3, -, 5, 6, 7, 8
Sorceror sta     : 1, 2, -, 4, 5, -, 7, 8
Sorceror har     : 1, -, 3, -, 5, -, 7, 8
Sorceror bal+sta : 1, 2, -, 4, 5, -, 7, 8
Sorceror bal+har : 1, -, 3, 4, -, 6, 7, -
Sorceror sta+har : 1, -, 3, 4, -, 6, 7, -
Sorceror zen     : 1, -, 3, -, 5, -, 7, -
```

**Comment by julealgon** [comment]
@AJenbo , I believe you meant "hit recovery" here and not "block". Please confirm and rename the issue accordingly.

**Comment by AJenbo** [comment]
i dont know the difference, but the function is PM_DoGotHit

**Comment by julealgon** [comment]
> i dont know the difference

Blocking is with shields, when you don't take damage. Hit recovery is when you take damage and are temporarily stunned.

> but the function is PM_DoGotHit

That's hit recovery (already failed armor and blocking checks at that point).

**Comment by StephenCWills** [comment]
A few notes from what I've done so far.

* Frame distributions were provided for Warrior, Rogue, and Sorcerer, but there's nothing for Monk, Bard, or Barbarian. I assume that because these additional classes have 6-, 7, and 6-frame animations respectively and so the corresponding distribution should apply to them as well.
* Although the game data only defines hit-recovery animations with 6, 7, or 8 frames, the original algorithm for skipping frames is technically generic for any number of frames. That said, I couldn't think of any generic way of achieving the exact distribution of frames listed in @AJenbo's original post. I decided the easiest thing was to simply hardcode the logic for 6-, 7-, and 8-frame animations and leave the original logic in as a fallback. If someone down the line really wants to add a character with 5-frame or 9-frame animations, they'll get the original logic unless they choose to extend the hardcoded logic.
* Unless I'm mistaken, because `ProcessPlayers()` is one of the first things that happens in `game_logic()`, it's not possible to skip the first frame of the animation by modifying `PM_DoGotHit()` alone. It should also be necessary to modify `StartPlrHit()`, which is called from a bunch of the other processes that occur later in `game_logic()`. However, it looks like this may break the test function in `player_test.cpp`. This is relevant for the Warrior's zen animation.

I have something implemented, but not yet tested. I'm going to develop a test for all these combinations. In the meantime, if anyone has any additional guidance after reading this, I would appreciate it.

**Comment by julealgon** [comment]
> Unless I'm mistaken, because ProcessPlayers() is one of the first things that happens in game_logic(), it's not possible to skip the first frame of the animation by modifying PM_DoGotHit()

@StephenCWills you might want to take a look at this (and the associated issues):
- https://github.com/diasurgical/devilutionX/pull/840

**Comment by StephenCWills** [comment]
@julealgon Thanks, I've reviewed that pull request and the associated issues. It appears to me that the `StartPlrHit()` function is nearly always called from code paths rooted in `ProcessMonsters()` or `ProcessMissiles()`. Correct me if I'm wrong, but this should mean that the first hit-recovery frame would not normally be skipped because `ProcessPlayers()` has completed before either of those other two functions are called. The only time it might be skipped is if the animation is triggered by the `PlrHitPlr()` function, and that really depends on the order in which players are processed in `ProcessPlayers()`.

I am a little curious about why the pull request was labeled wontfix, but that's almost certainly unrelated to this issue.

**Comment by AJenbo** [comment]
Because it changes game balance by slowing down the attack rate. Like with block animations, a lot of deliberate game balance was done by simply skipping frames so changing it would alter that balance, even if it may essentially be a bug.

> I am a little curious about why the pull request was labeled wontfix, but that's almost certainly unrelated to this issue.

**Comment by StephenCWills** [comment]
> Because it changes game balance by slowing down the attack rate. Like with block animations, a lot of deliberate game balance was done by simply skipping frames so changing it would alter that balance, even if it may essentially be a bug.

Thanks for the explanation. In this case, the first hit-recovery frame could be skipped if player 1 hits player 2, but the change in animation state occurs when processing player 1's attack. Therefore, `PM_DoGotHit()` will still be called on the first frame for player 2 regardless of whether the frame is skipped. This behavior shouldn't change as a result of modifying the distribution of skipped frames so game balance won't be affected.

In terms of the bigger picture, it seems to me that the skipped hit-recovery frame would give melee attackers a slight disadvantage in PvP if they join the game in an earlier slot than their opponent. Given the conditions by which the frame is skipped, I imagine it was not a deliberate game-balancing decision. However, any fix would probably be more complex than it's worth, but that's just my perspective as an avid PvE'er. Also, the fix proposed in #840 would actually detrimentally affect the hit-recovery animations by adding a delay on the first frame transition in most cases.

I'll refrain from commenting here on any other animations besides hit-recovery to avoid derailing the conversation any further.

**Comment by julealgon** [comment]
> Correct me if I'm wrong, but this should mean that the first hit-recovery frame would not normally be skipped because ProcessPlayers()

That's correct. When the action is triggered from a point in time after `ProcessPlayers` has already executed, the first frame is shown normally. This is why hit recovery is not affected.

> Thanks for the explanation. In this case, the first hit-recovery frame could be skipped if player 1 hits player 2, but the change in animation state occurs when processing player 1's attack.

I had not considered the implications of multiplayer when looking at it. Your observation makes sense to me.

> However, any fix would probably be more complex than it's worth

Agreed, not with the "worthiness" part of it, but the complexity. I definitely think it would be worth it, but the work to get there would be significant. We'd probably have to completely decouple animation from logic and run _all_ animations after _all_ logic or something along those lines to avoid the interdependency of actions.

> Also, the fix proposed in #840 would actually detrimentally affect the hit-recovery animations by adding a delay on the first frame transition in most cases.

Hmm interesting... I assume you are correct. I'll check that again later (even though it's been stale for a while).

**Comment by StephenCWills** [comment]
@AJenbo, FYI, I just noticed a discrepancy between the the following frame distributions.

```
Sorceror har     : 1, -, 3, -, 5, -, 7, 8
Sorceror bal+har : 1, -, 3, 4, -, 6, 7, -
Sorceror sta+har : 1, -, 3, 4, -, 6, 7, -
```

The logic I defined in my pull request actually just uses `1, -, 3, -, 5, -, 7, 8` for all three of these. I didn't write any code to differentiate between these combinations so hopefully this is fine.

**Comment by StephenCWills** [comment]
I just discovered this issue is still open. I believe this was resolved with #1468 so I'll just go ahead and close it.
